watertap/property_models/ion_DSPMDE_prop_pack.py

```diff
@@ -640,21 +790,68 @@ def define_state_vars(self):
+                                                 for j in self.params.solute_set) == 0)
+            else:
+                raise ValueError('adjust_by_ion must be set to the name of an ion in the list of solutes.')
+        if defined_state:
             for j in self.params.solute_set:
-                if not self.flow_mol_phase_comp['Liq', j].is_fixed():
+                if not self.flow_mol_phase_comp['Liq', j].is_fixed() and adjust_by_ion != j:
                     raise AssertionError(
                         f"{self.flow_mol_phase_comp['Liq', j]} was not fixed. Fix flow_mol_phase_comp for each solute"
                         f" to check that electroneutrality is satisfied.")
-        val = value(sum(self.charge_comp[j] * self.flow_mol_phase_comp['Liq', j]
-                    for j in self.params.solute_set))
+                if adjust_by_ion == j and self.flow_mol_phase_comp['Liq', j].is_fixed():
+                    self.flow_mol_phase_comp['Liq', j].unfix()
+        else:
+            for j in self.params.solute_set:
+                if self.flow_mol_phase_comp['Liq', j].is_fixed():
+                    raise AssertionError(
+                        f"{self.flow_mol_phase_comp['Liq', j]} was fixed. Either set defined_state=True or unfix "
+                        f"flow_mol_phase_comp for each solute to check that electroneutrality is satisfied.")
+
+        if get_property is not None:
+            for i in [get_property]:
+                getattr(self, i)
+        self.conc_mol_phase_comp
+
+        if solve:
+            solve = get_solver()
+            results = solve.solve(self)
+            if check_optimal_termination(results):
+                val = value(sum(self.charge_comp[j] * self.conc_mol_phase_comp['Liq', j]
+                            for j in self.params.solute_set))
+            else:
+                if adjust_by_ion is not None:
+                    del self.charge_balance
+                raise ValueError('The stateblock failed to solve while computing concentrations to check the charge balance.')
+        else:
+            val = value(sum(self.charge_comp[j] * self.conc_mol_phase_comp['Liq', j]
+                            for j in self.params.solute_set))
         if abs(val) <= tol:
+            if adjust_by_ion is not None:
+                # self.flow_mol_phase_comp['Liq', adjust_by_ion].fix()
+                ion_adjusted = self.flow_mol_phase_comp['Liq', adjust_by_ion].value
+                if defined_state:
+                    self.flow_mol_phase_comp['Liq', adjust_by_ion].fix(ion_adjusted)
+                    msg = f"{adjust_by_ion} was adjusted and flow_mol_phase_comp['Liq',{adjust_by_ion}] was fixed " \
+                          f"to {ion_adjusted}."
+                else:
+                    msg = f"{adjust_by_ion} was adjusted and the value computed for flow_mol_phase_comp['Liq',{adjust_by_ion}]" \
+                          f" is {ion_adjusted}."
+                del self.charge_balance
```

Comment:
It's hard to verify that this constraint gets deleted in every case, and if the logic of this method changes there's the potential for bugs, even if none exist presently.

Maybe we should just check for its existence at the end of the method and delete it if it's still there?